### PR TITLE
Added the coordinate descent method from scikit-learn, supporting sparsity

### PR DESCRIPTION
**Pull request by vilim**
This is a more-or-less direct translation of the Python Scikit-learn NMF method, which hopefully respects Julian idioms.

I am not sure how I should reference the original implementation, so any comments on attribution, licences and copyright are welcome.

### CODE REVIEW

**Comment by kmsquire** [comment]
[Here](https://github.com/JuliaImages/Images.jl/blob/812ac49cf5c5f03c3b26423dbb43d9729aaa954d/src/edge.jl#L157)'s an example of some publicly available code that was translated from Matlab and placed in Images.jl.

Scikit-learn code is available under a BSD-style license.  This package is MIT-licensed.  Generally, these licenses are compatible with each other, so as long as the core maintainers of this package are open to it, a typical solution would be to note an exception in `LICENSE.md`, which states that the package is available under the MIT license (already there), except as noted at the top of individual files, and then to include the copyright and BSD license text at the top of `coorddesc.jl`.

As long as you include the copyright and license somewhere, any variation on above should be fine.

**Comment by ararslan** [comment]
> except as noted at the top of individual files

Note that BSD style licenses are not copyleft, so we don't need to have an exception in our license for code derived from BSD-licensed code; BSD style licenses permit licensing derived works under any license. We can simply acknowledge and link to the license for the code from which this was derived directly in the file.

**Comment by vilim** [comment]
@ararslan I have already updated the file referencing the original implementation and the license, is there anything else to do before it can be merged? The same Travis test error is also present on the master branch.

File: Project.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+name = "NMF"
+uuid = "6ef6ca0d-6ad7-5ff6-b225-e928bfa0a386"
+license = "MIT"
+version = "0.4.1"
```
Comment (ararslan) [inline Project.toml]: Please remove the version. It just gets annoying to keep it in sync with the tags and it isn't actually necessary.

File: Project.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+[deps]
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Comment (ararslan) [inline Project.toml]: Test should actually be a test dependency, which lives in its own section.
```toml
[extras]
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["Test"]
```
Any other packages which are used in testing but not in the package should be under `[extras]` and similarly listed in the `test` target.

File: Manifest.toml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (ararslan) [inline Manifest.toml]: Generally speaking, Manifest.toml files shouldn't be committed. Please `git rm` this and instead add `Manifest.toml` to the .gitignore file.

File: src/coorddesc.jl [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+# Julia translation: Vilim Štih
+
+# Reference: Cichocki, Andrzej, and P. H. A. N. Anh-Huy. "Fast local algorithms for
+#  large scale nonnegative matrix and tensor factorizations."
+#  IEICE transactions on fundamentals of electronics, communications and
+#  computer sciences 92.3: 708-721, 2009.
+
+
+mutable struct TransposedArray{T} <: AbstractArray{T,2}
```
Comment (ararslan) [inline src/coorddesc.jl]: Is this actually necessary? Starting with Julia 0.7, `transpose` and `adjoint` produce lazy wrappers of the underlying matrix, which I think should fulfill this purpose.

File: src/coorddesc.jl [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    A :: AbstractArray{T, 2}
+end
+
+Base.size(x::TransposedArray) = size(x.A)[end:-1:1]
+Base.getindex(x::TransposedArray, I::Vararg{Int, 2}) = fybc
```
Comment (ararslan) [inline src/coorddesc.jl]: `fybc`?

File: src/coorddesc.jl [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            end
+        end
+    end
+    return violation
+end
+
+
+function update_wh!(upd::CoordinateDescentUpd{T}, s::CoordinateDescentState{T},
+     X::AbstractArray{T}, W::AbstractArray{T}, H::AbstractArray{T}) where T
+     Ht = PermutedDimsArray(H, (2, 1))
```
Comment (ararslan) [inline src/coorddesc.jl]: Looks like this down to line 147 have 5 spaces for indentation instead of 4.

**Comment by ararslan** [comment]
> I have already updated the file referencing the original implementation and the license

I saw, that looks good. My comment about the license was in response to Kevin's.

File: src/coorddesc.jl (hunk 4 above)
Comment (vilim) [reply src/coorddesc.jl]: Thanks for catching this. I was actually using permuteddimsarray, which has now been updated to use the lazy transpose.

**Comment by vilim** [comment]
@ararslan Thanks for the review! I have implemented all the changes.

**Comment by ararslan** [comment]
Looks like the Project.toml changes weren't implemented.

**Comment by vilim** [comment]
@ararslan sorry, I switched to VS Code recently and am still getting used to file status indicators; it wasn't committed. It is fixed now.

**Comment by vilim** [comment]
@ararslan is there anything else I need to do before this can be merged?

File: src/interf.jl [hunk 7]
```diff
@@ -35,6 +35,8 @@ function nnmf(X::AbstractMatrix{T}, k::Integer;
         alginst = MultUpdate{T}(obj=:mse, maxiter=maxiter, tol=tol, verbose=verbose)
     elseif alg == :multdiv
         alginst = MultUpdate{T}(obj=:div, maxiter=maxiter, tol=tol, verbose=verbose)
+    elseif alg == :cd
```
Comment (ararslan) [inline src/interf.jl]: Can you add a test for calling `nnmf` with `alg=:cd`?

Comment (vilim) [reply src/interf.jl]: I added a test (revealing a bug, thanks!), also added interf.jl test to the list of tests performed. This in turn revealed a deprecation in projals.jl (vecnorm to norm) which I also had to fix in order for the tests to pass.

**Comment by VPetukhov** [comment]
Hi! Can I ask what's blocking from merging this PR? Looks like the requested changes were resolved. And I can't reproduce the test failure: now it works absolutely fine. Here is travis the successful report from my fork (without any changes): https://travis-ci.com/VPetukhov/NMF.jl/builds/120496869 . So can you please maybe trigger rebuild and check if it still fails?

**Comment by VPetukhov** [comment]
Thank you!

**Pull request merged**